File: app/src/main/cpp/egl/TXEglHelp.cpp
/**
 * egl 初始化
 * Created by yin13 on 2021/6/12
 * Email:yin13753884368@163.com
 * CSDN:http://blog.csdn.net/yin13753884368/article
 * Github:https://github.com/taxiao213
*/

#include "TXEglHelp.h"

TXEglHelp::TXEglHelp() {
    this->mEGLDisplay = EGL_NO_DISPLAY;
    this->mEGLContext = EGL_NO_CONTEXT;
    this->mEGLSurface = EGL_NO_SURFACE;
    this->mEGLConfig = NULL;
}

TXEglHelp::~TXEglHelp() {


}

int TXEglHelp::initEgl(EGLNativeWindowType windowType) {
    // 1.初始化 EGLDisplay
    SDK_LOG_D("初始化 EGLDisplay");
    mEGLDisplay = eglGetDisplay(EGL_DEFAULT_DISPLAY);
    if (mEGLDisplay == EGL_NO_DISPLAY) {
        SDK_LOG_E("初始化失败 EGLDisplay");
        return -1;
    }

    // 2.初始化 eglInitialize
    SDK_LOG_D("初始化 eglInitialize");
    EGLint *egLint = new EGLint(2);
    if (!eglInitialize(mEGLDisplay, &egLint[0], &egLint[1])) {
        SDK_LOG_E("初始化失败 eglInitialize");
        return -1;
    }

    // 3.设置显示的相关属性
    // 占用一个字节 8 bit, 以 EGL_NONE 结尾, 每次获取 2 位数值
    // 颜色 R G B ，depth 深度, stencil 模板等级
    SDK_LOG_D("初始化 设置显示的相关属性");
    const EGLint attrib[] = {
            EGL_RED_SIZE, 8,
            EGL_GREEN_SIZE, 8,
            EGL_BLUE_SIZE, 8,
            EGL_ALPHA_SIZE, 8,
            EGL_DEPTH_SIZE, 8,
            EGL_STENCIL_SIZE, 8,
            EGL_RENDERABLE_TYPE, EGL_OPENGL_ES2_BIT,
            EGL_NONE
    };

    // 4.初始化 eglChooseConfig
    SDK_LOG_D("初始化 eglChooseConfig");
    EGLint num_config;
    // 传 1 只获取一个 config，传 null 在某些设备可能获取不到值
    if (!eglChooseConfig(mEGLDisplay, attrib, NULL, 1, &num_config)) {
        SDK_LOG_E("初始化失败 eglChooseConfig1");
        return -1;
    }
    // 获取到的 num_config 传进去，获取到 EGLConfig
    if (!eglChooseConfig(mEGLDisplay, attrib, &mEGLConfig, num_config, &num_config)) {
        SDK_LOG_E("初始化失败 eglChooseConfig2");
        return -1;
    }

    // 5.初始化 eglCreateContext
    // CLIENT_VERSION egl版本 ，这里用的是 2 版本
    SDK_LOG_D("初始化 eglCreateContext");
    EGLint attrib_list[] = {
            EGL_CONTEXT_CLIENT_VERSION, 2,
            EGL_NONE
    };
    mEGLContext = eglCreateContext(mEGLDisplay, mEGLConfig, EGL_NO_CONTEXT, attrib_list);
    if (mEGLContext == NULL) {
        SDK_LOG_E("初始化失败 eglCreateContext");
        return -1;
    }

    // 6.初始化 eglCreateWindowSurface
    SDK_LOG_D("初始化 eglCreateWindowSurface");
    mEGLSurface = eglCreateWindowSurface(mEGLDisplay, mEGLConfig, windowType, NULL);
    if (mEGLSurface == NULL) {
        SDK_LOG_E("初始化失败 eglCreateWindowSurface");
        return -1;
    }

    // 7.初始化 eglMakeCurrent 绑定EglContext和Surface到显示设备中
    SDK_LOG_D("初始化 eglMakeCurrent");
    if (!eglMakeCurrent(mEGLDisplay, mEGLSurface, mEGLSurface, mEGLContext)) {
        SDK_LOG_E("初始化失败 eglMakeCurrent");
        return -1;
    }

    SDK_LOG_D("egl init success! ");
    return 0;
}

int TXEglHelp::swapBuffers() {
    if (mEGLDisplay == NULL || mEGLSurface == NULL || mEGLDisplay == EGL_NO_DISPLAY ||
        mEGLSurface == EGL_NO_SURFACE) {
        SDK_LOG_E("mEGLDisplay 为 null , mEGLSurface  为 null ");
        return -1;
    }
    // 8. 初始化 eglSwapBuffers 刷新数据 显示渲染场景
    if (!eglSwapBuffers(mEGLDisplay, mEGLSurface)) {
        SDK_LOG_E("初始化失败 eglSwapBuffers");
        return -1;
    }
    return 0;
}

int TXEglHelp::destroyEgl() {

    if (mEGLDisplay == NULL || mEGLSurface == NULL || mEGLDisplay == EGL_NO_DISPLAY ||
        mEGLSurface == EGL_NO_SURFACE) {
        SDK_LOG_E("mEGLDisplay 为 null , mEGLSurface  为 null ");
        return -1;
    }
    if (mEGLDisplay != EGL_NO_DISPLAY) {
        SDK_LOG_E("destroyEgl eglMakeCurrent ");
        eglMakeCurrent(mEGLDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
    }

    if (mEGLDisplay != EGL_NO_DISPLAY && mEGLSurface != EGL_NO_SURFACE) {
        SDK_LOG_E("destroyEgl eglDestroySurface ");
        eglDestroySurface(mEGLDisplay, mEGLSurface);
        mEGLSurface = EGL_NO_SURFACE;
    }

    if (mEGLDisplay != EGL_NO_DISPLAY && mEGLContext != EGL_NO_CONTEXT) {
        SDK_LOG_E("destroyEgl eglDestroyContext ");
        eglDestroyContext(mEGLDisplay, mEGLContext);
        mEGLContext = EGL_NO_CONTEXT;
    }

    if (mEGLDisplay != EGL_NO_DISPLAY) {
        SDK_LOG_E("destroyEgl eglTerminate ");
        eglTerminate(mEGLDisplay);
        mEGLDisplay = EGL_NO_DISPLAY;
    }
    return 0;
}

File: app/src/main/java/com/taxiao/opengl/util/egl/TXEglHelp.java
package com.taxiao.opengl.util.egl;

import android.opengl.EGL14;
import android.util.Log;
import android.view.Surface;

import javax.microedition.khronos.egl.EGL10;
import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.egl.EGLContext;
import javax.microedition.khronos.egl.EGLDisplay;
import javax.microedition.khronos.egl.EGLSurface;


/**
 * Egl 环境
 * Created by hanqq on 2021/6/23
 * Email:yin13753884368@163.com
 * CSDN:http://blog.csdn.net/yin13753884368/article
 * Github:https://github.com/taxiao213
 */
public class TXEglHelp {
    private static String TAG = TXEglHelp.class.getName();
    private EGL10 mEgl;
    private EGLDisplay mEglDisplay;
    private EGLContext mEglContext;
    private EGLSurface mEglSurface;

    public void init(Surface surface, EGLContext eglContext) {
        Log.d(TAG, "createEngine");
        // 1.Get an EGL instance
        mEgl = (EGL10) EGLContext.getEGL();

        // 2.Get to the default display.
        mEglDisplay = mEgl.eglGetDisplay(EGL10.EGL_DEFAULT_DISPLAY);

        if (mEglDisplay == EGL10.EGL_NO_DISPLAY) {
            throw new RuntimeException("eglGetDisplay failed");
        }

        // 3.We can now initialize EGL for that display
        int[] version = new int[2];
        if (!mEgl.eglInitialize(mEglDisplay, version)) {
            throw new RuntimeException("eglInitialize failed");
        }

        // 4.创建属性
        int[] attrbutes = new int[]{
                EGL10.EGL_RED_SIZE, 8,
                EGL10.EGL_GREEN_SIZE, 8,
                EGL10.EGL_BLUE_SIZE, 8,
                EGL10.EGL_ALPHA_SIZE, 8,
                EGL10.EGL_DEPTH_SIZE, 8,
                EGL10.EGL_STENCIL_SIZE, 8,
                EGL10.EGL_RENDERABLE_TYPE, EGL14.EGL_OPENGL_ES2_BIT,
                EGL10.EGL_NONE
        };

        // 5.获取 config
        int[] num_config = new int[1];
        if (!mEgl.eglChooseConfig(mEglDisplay, attrbutes, null, 0,
                num_config)) {
            throw new IllegalArgumentException("eglChooseConfig failed");
        }

        int numConfigs = num_config[0];

        if (numConfigs <= 0) {
            throw new IllegalArgumentException(
                    "No configs match configSpec");
        }

        EGLConfig[] configs = new EGLConfig[numConfigs];
        if (!mEgl.eglChooseConfig(mEglDisplay, attrbutes, configs, numConfigs,
                num_config)) {
            throw new IllegalArgumentException("eglChooseConfig#2 failed");
        }

        // 6.获取 EGLContext
        int[] attrib_list = {
                EGL14.EGL_CONTEXT_CLIENT_VERSION,
                2,
                EGL10.EGL_NONE
        };

        if (eglContext != null) {
            Log.d(TAG, "eglContext != null");
            mEglContext = mEgl.eglCreateContext(mEglDisplay, configs[0], eglContext, attrib_list);
        } else {
            Log.d(TAG, "eglContext == null");
            mEglContext = mEgl.eglCreateContext(mEglDisplay, configs[0], EGL10.EGL_NO_CONTEXT, attrib_list);
        }

        // 7.CreateWindowSurface
        mEglSurface = mEgl.eglCreateWindowSurface(mEglDisplay, configs[0], surface, null);

        if (!mEgl.eglMakeCurrent(mEglDisplay, mEglSurface, mEglSurface, mEglContext)) {
            throw new RuntimeException("eglMakeCurrent fail");
        }
    }

    public void swapBuffers() {
        if (mEglDisplay == null || mEglSurface == null || mEglDisplay == EGL10.EGL_NO_DISPLAY ||
                mEglSurface == EGL10.EGL_NO_SURFACE) {
            Log.d(TAG, "mEGLDisplay 为 null , mEGLSurface  为 null ");
            return;
        }
        // 8. 初始化 eglSwapBuffers 刷新数据 显示渲染场景
        if (!mEgl.eglSwapBuffers(mEglDisplay, mEglSurface)) {
            Log.d(TAG, "初始化失败 eglSwapBuffers");
            return;
        }
        Log.d(TAG, "success eglSwapBuffers");
    }

    public EGLContext getmEglContext() {
        return mEglContext;
    }

    public void destoryEgl() {
        if (mEgl != null) {
            mEgl.eglMakeCurrent(mEglDisplay, EGL10.EGL_NO_SURFACE,
                    EGL10.EGL_NO_SURFACE,
                    EGL10.EGL_NO_CONTEXT);

            mEgl.eglDestroySurface(mEglDisplay, mEglSurface);
            mEglSurface = null;

            mEgl.eglDestroyContext(mEglDisplay, mEglContext);
            mEglContext = null;

            mEgl.eglTerminate(mEglDisplay);
            mEglDisplay = null;
            mEgl = null;
        }
    }
}
